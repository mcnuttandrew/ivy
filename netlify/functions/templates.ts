import {getMany} from '../utils';

export const handler = getMany('templates');
