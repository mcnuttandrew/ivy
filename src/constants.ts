// i know this file is bad and crazy i feel bad
export interface OptionValue {
  display: string;
  value: string;
}
export interface EncodingOption {
  optionType: string;
  options: OptionValue[];
  optionSetter: (spec: any, selectedOption: string) => any;
  optionGetter: (spec: any) => string;
  optionDefault: string;
  predicate: (spec: any) => boolean;
}

import {VegaMark} from './types';
export const marks: VegaMark[] = [
  'arc',
  'area',
  // 'image',
  // 'group',
  // 'line',
  // 'path',
  'rect',
  // 'rule',
  // 'shape',
  'symbol',
  'text',
  // 'trail',
];

// check if using the nested spec mode, if so, re-route into that subspec
const maybePrefixWithSpec = (spec: any, route: string[]): string[] =>
  Boolean(spec.getIn(['spec'])) ? ['spec', ...route] : route;

const justCountAgg = [
  {display: 'none', value: undefined},
  {display: 'count', value: 'count'},
];
const spatialAggs = [
  ...justCountAgg,
  {display: 'min', value: 'min'},
  {display: 'max', value: 'max'},
  {display: 'sum', value: 'sum'},
  {display: 'mean', value: 'mean'},
  {display: 'median', value: 'median'},
  {display: 'mode', value: 'mode'},
];
const toOption = (x: string): OptionValue => ({display: x, value: x});
const binningOptions = [
  'none',
  'count',
  'min',
  'mean',
  'sum',
  'bin', // TODO bin is handled incorrectly
  'max',
  'median',
].map(toOption);

const buildSpatialOptions = (
  dimension: string,
  options: OptionValue[],
): EncodingOption => ({
  optionType: 'Agg.',
  options,
  optionSetter: (spec, selectedOption) => {
    const route = maybePrefixWithSpec(spec, ['encoding', dimension]);
    const newSpec = spec.setIn([...route, 'aggregate'], selectedOption);
    const addType = (x: any) => x.setIn([...route, 'type'], 'quantitative');
    const hasField = newSpec.getIn([...route, 'field']);
    return hasField ? newSpec : addType(newSpec);
  },
  optionGetter: spec => {
    const route = maybePrefixWithSpec(spec, [
      'encoding',
      dimension,
      'aggregate',
    ]);
    return spec.getIn(route);
  },
  optionDefault: 'none',
  predicate: () => true,
});

const typeRoute = (dim: string) => ['encoding', dim, 'scale', 'type'];
const buildScaleOption = (dim: string): EncodingOption => ({
  optionType: 'Scale type',
  options: ['linear', 'log'].map(toOption),
  optionSetter: (spec, option) =>
    spec.setIn(maybePrefixWithSpec(spec, typeRoute(dim)), option),
  optionGetter: spec =>
    spec.getIn(maybePrefixWithSpec(spec, typeRoute(dim))) || 'linear',
  optionDefault: 'linear',
  predicate: spec =>
    Boolean(spec.getIn(maybePrefixWithSpec(spec, ['encoding', dim, 'field']))),
});

const buildCoerceRoute = (spec: any, dim: string): string[] =>
  maybePrefixWithSpec(spec, ['encoding', dim, 'type']);
const buildTypeCoercion = (dim: string): EncodingOption => ({
  optionType: 'Data type',
  options: ['nominal', 'ordinal', 'quantitative', 'temporal'].map(toOption),
  optionSetter: (spec, option) =>
    spec.setIn(buildCoerceRoute(spec, dim), option),
  optionGetter: spec => spec.getIn(buildCoerceRoute(spec, dim)),
  optionDefault: undefined,
  predicate: () => true,
});

// take in an encoding option and create a predicate that rejects that option
// if that field is not present
type predicateInject = (dim: string, option: EncodingOption) => EncodingOption;
const buildPredRoute = (spec: any, dim: string): string[] =>
  maybePrefixWithSpec(spec, ['encoding', dim, 'field']);
const injectFieldPred: predicateInject = (dim, option) => ({
  ...option,
  predicate: (spec: any): boolean => !!spec.getIn(buildPredRoute(spec, dim)),
});
const injectNofieldPred: predicateInject = (dim, option) => ({
  ...option,
  predicate: (spec: any): boolean => !spec.getIn(buildPredRoute(spec, dim)),
});

const generateXorY = (dim: string) => [
  injectFieldPred(dim, buildTypeCoercion(dim)),
  buildScaleOption(dim),
  injectFieldPred(dim, buildSpatialOptions(dim, spatialAggs)),
  injectNofieldPred(dim, buildSpatialOptions(dim, justCountAgg)),
];

// the configuration options consumed by the encoding panel
export const configurationOptions: any = {
  x: generateXorY('x'),
  y: generateXorY('y'),
  'row-meta': [],
  'column-meta': [],
  'repeat-meta': [],

  // repeaters
  facet: [injectFieldPred('facet', buildTypeCoercion('facet'))],
  row: [injectFieldPred('row', buildTypeCoercion('row'))],
  column: [injectFieldPred('column', buildTypeCoercion('column'))],
  // marks
  size: [
    injectFieldPred('size', buildTypeCoercion('size')),
    injectFieldPred('size', buildSpatialOptions('size', binningOptions)),
    injectNofieldPred('size', buildSpatialOptions('size', justCountAgg)),
  ],
  color: [
    injectFieldPred('color', buildTypeCoercion('color')),
    injectFieldPred('color', buildSpatialOptions('color', binningOptions)),
    injectNofieldPred('color', buildSpatialOptions('color', justCountAgg)),
  ],
  shape: [
    injectFieldPred('shape', buildTypeCoercion('shape')),
    injectFieldPred('shape', buildSpatialOptions('shape', binningOptions)),
  ],
  detail: [
    injectFieldPred('detail', buildTypeCoercion('detail')),
    injectFieldPred('detail', buildSpatialOptions('detail', binningOptions)),
  ],
  text: [
    injectFieldPred('text', buildTypeCoercion('text')),
    injectFieldPred('text', buildSpatialOptions('text', binningOptions)),
    injectNofieldPred('text', buildSpatialOptions('text', justCountAgg)),
  ],
};
