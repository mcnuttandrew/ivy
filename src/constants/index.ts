export const test = 'test';
type ON = 'on';
export const EDITOR_OPTIONS = {
  selectOnLineNumbers: true,
  automaticLayout: true,
  wordWrap: 'on' as ON,
  minimap: {
    enabled: false,
  },
  fontSize: 10,
  lineDecorationsWidth: 0,
  lineNumbersMinChars: 3,
};

export const TEXT_TYPE: {[x: string]: string} = {
  MEASURE: '123',
  DIMENSION: 'ABC',
  TIME: 'ðŸ•‘',
};

export const GRAMMAR_NAME = 'grammer';
export const GRAMMAR_DESC = 'Tableau-style grammar of graphics';

// export const
export const WIDGET_VALUES = 'Settings';
export const WIDGET_CONFIGURATION = 'Params';
export const JSON_OUTPUT = 'Output';
export const TEMPLATE_BODY = 'Body';

/* eslint-disable no-undef*/
export const PREVENT_ACCIDENTAL_LEAVE = process.env.NODE_ENV === 'production';
/* eslint-enable no-undef*/
