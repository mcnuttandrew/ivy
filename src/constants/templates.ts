import {DataType} from '../types';
// probably need to have a notion of typing for some of the widgets
export type WidgetType = 'DataTarget' | 'List' | 'Switch' | 'Text';
export interface TemplateWidget {
  widgetName: string;
  widgetType: WidgetType;
  required: boolean;
}
export interface DataTargetWidget extends TemplateWidget {
  widgetType: 'DataTarget';
  allowedTypes: DataType[];
}
export interface ListWidget extends TemplateWidget {
  widgetType: 'List';
  allowedValues: string[];
  defaultValue: string;
}
export interface SwitchWidget extends TemplateWidget {
  widgetType: 'Switch';
  defaultValue: boolean;
}
export interface TextWidget extends TemplateWidget {
  widgetType: 'Text';
  text: string;
}

export interface Template {
  templateName: string;
  code: string;
  widgets: (
    | TemplateWidget
    | DataTargetWidget
    | ListWidget
    | SwitchWidget
    | TextWidget
  )[];
  // TODO MAYBE ADD A PREVIEW PIC?
}

export interface TemplateMap {
  [key: string]: string;
}

const SCATTERPLOT_EXAMPLE: any = {
  mark: {
    type: 'point',
    tooltip: true,
  },
  encoding: {
    x: {
      field: '[xDim]',
      type: 'quantitative',
    },
    y: {
      field: '[yDim]',
      type: 'quantitative',
    },
  },
};

const SCATTERPLOT_TEMPLATE: Template = {
  templateName: 'scatterplot',
  code: JSON.stringify(SCATTERPLOT_EXAMPLE, null, 2),
  widgets: [
    {
      widgetName: 'xDim',
      widgetType: 'DataTarget',
      allowedTypes: ['MEASURE'],
      required: true,
    },
    {
      widgetName: 'yDim',
      widgetType: 'DataTarget',
      allowedTypes: ['MEASURE'],
      required: true,
    },
  ],
};

export const DEFAULT_TEMPLATES: Template[] = [SCATTERPLOT_TEMPLATE];
