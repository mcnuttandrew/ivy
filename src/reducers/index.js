import {createStore, combineReducers, applyMiddleware} from 'redux';
import thunk from 'redux-thunk';
import Immutable, {Map} from 'immutable';

const DEFAULT_STATE = Immutable.fromJS({});

const actionFuncMap = {};
const NULL_ACTION = (state, payload) => state;

export default createStore(
  combineReducers({
    base: (state = DEFAULT_STATE, {type, payload}) => {
      return (actionFuncMap[type] || NULL_ACTION)(state, payload);
    },
  }),
  applyMiddleware(thunk),
);
