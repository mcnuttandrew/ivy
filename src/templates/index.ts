import {
  DataTargetWidget,
  DataType,
  FreeTextWidget,
  GenWidget,
  ListWidget,
  MultiDataTargetWidget,
  SectionWidget,
  SliderWidget,
  SwitchWidget,
  Template,
  TextWidget,
  Widget,
} from '../types';
import {toList} from '../utils';
import {VEGA_CATEGORICAL_COLOR_SCHEMES, tableau10} from './vega-common';
import ATOM from './atom';
import BEESWARM_TEMPLATE from './bee-swarm';
import DATATABLE from './table';
import PIECHART_TEMPLATE from './pie-chart';
import SCATTERPLOT_TEMPLATE from './scatterplot';
import POLESTAR from './polestar-template';
import UNITVIS from './atom-example';
import PARALLEL_COORDINATES from './parallel-coordinates';
import RADAR from './radar-chart';

const DATA_TYPES: DataType[] = ['MEASURE', 'DIMENSION', 'TIME'];
export type WidgetFactoryFunc = (name: number | string) => GenWidget;
const makeName = (name: string | number, prefix: string): string =>
  typeof name === 'string' ? name : `${prefix}${name}`;
export const WidgetDescriptions: {[x: string]: string} = {
  DataTarget: 'For single fields from the current data set.',
  MultiDataTarget: 'For arrays of fields from the current data set.',
  List: 'For selecting one of several specific values.',
  Switch: 'For toggling between two specific values.',
  Slider: 'For specifying a particular numeric value.',
  FreeText: 'For specifying a free text.',
  Section: 'For organizing widgets into understandable sections.',
  Text: 'For explaining the purpose of widgets within the UI.',
};
export const DataTargetFactory: WidgetFactoryFunc = idx =>
  ({
    name: makeName(idx, 'Var'),
    type: 'DataTarget',
    config: {allowedTypes: DATA_TYPES, required: true},
  } as Widget<DataTargetWidget>);
export const MultiDataTargetFactory: WidgetFactoryFunc = idx =>
  ({
    name: makeName(idx, 'MultiDim'),
    type: 'MultiDataTarget',
    config: {allowedTypes: DATA_TYPES, required: true, minNumberOfTargets: 0},
  } as Widget<MultiDataTargetWidget>);
export const ListFactory: WidgetFactoryFunc = idx =>
  ({
    name: makeName(idx, 'ListItem'),
    type: 'List',
    config: {allowedValues: [] as {display: string; value: string}[], defaultValue: null},
  } as Widget<ListWidget>);
export const SwitchFactory: WidgetFactoryFunc = idx =>
  ({
    name: makeName(idx, 'Switch'),
    type: 'Switch',
    config: {active: 'true', inactive: 'false', defaultsToActive: true},
  } as Widget<SwitchWidget>);
export const SliderFactory: WidgetFactoryFunc = idx =>
  ({
    name: `Slider${idx}`,
    type: 'Slider',
    config: {minVal: 0, maxVal: 10, step: 1, defaultValue: 5},
  } as Widget<SliderWidget>);
export const SectionFactory: WidgetFactoryFunc = idx =>
  ({name: makeName(idx, 'Section'), type: 'Section', config: null} as Widget<SectionWidget>);
export const FreeTextFactory: WidgetFactoryFunc = idx =>
  ({name: makeName(idx, 'FreeText'), type: 'FreeText', config: {useParagraph: false}} as Widget<
    FreeTextWidget
  >);
export const TextFactory: WidgetFactoryFunc = idx =>
  ({name: makeName(idx, 'Text'), type: 'Text', config: {text: ''}} as Widget<TextWidget>);

export const widgetFactory: {[type: string]: WidgetFactoryFunc} = {
  DataTarget: DataTargetFactory,
  MultiDataTarget: MultiDataTargetFactory,
  List: ListFactory,
  Switch: SwitchFactory,
  Slider: SliderFactory,
  Section: SectionFactory,
  FreeText: FreeTextFactory,
  Text: TextFactory,
};

export const widgetFactoryByGroups: {[type: string]: {[x: string]: WidgetFactoryFunc}} = {
  'Data Widgets': {
    DataTarget: DataTargetFactory,
    MultiDataTarget: MultiDataTargetFactory,
  },
  'Value Widgets': {
    List: ListFactory,
    Switch: SwitchFactory,
    Slider: SliderFactory,
    FreeText: FreeTextFactory,
  },
  'Mark up Widgets': {
    Section: SectionFactory,
    Text: TextFactory,
    // Shortcuts: ShortcutsFactory,
  },
};

export const preconfiguredWidgets: {[type: string]: WidgetFactoryFunc} = {
  'Discrete Color Options': idx =>
    ({
      name: `ColorList${idx}`,
      type: 'List',
      config: {allowedValues: toList(VEGA_CATEGORICAL_COLOR_SCHEMES)},
    } as Widget<ListWidget>),
  'Data Types Options': idx =>
    ({
      name: `DataTypeOptions${idx}`,
      type: 'List',
      config: {allowedValues: toList(['quantitative', 'temporal', 'ordinal', 'nominal'])},
    } as Widget<ListWidget>),
  'Single Color Options (Tableau 10)': idx =>
    ({
      name: `DataTypeOptions${idx}`,
      type: 'List',
      config: {allowedValues: toList(tableau10)},
    } as Widget<ListWidget>),
};

export const DEFAULT_TEMPLATES: Template[] = [
  POLESTAR,
  ATOM,
  DATATABLE,
  SCATTERPLOT_TEMPLATE,
  PIECHART_TEMPLATE,
  BEESWARM_TEMPLATE,
  UNITVIS,
  PARALLEL_COORDINATES,
  RADAR,
];
