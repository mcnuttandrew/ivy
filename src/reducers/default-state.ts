import produce from 'immer';
import stringify from 'json-stringify-pretty-compact';
import {ColumnHeader} from '../types';
import {Template, TemplateMap} from '../templates/types';
// import SCATTERPLOT from '../templates/example-templates/scatterplot';
export interface UndoRedoStackItem {
  spec: any;
  currentView: string;
  templateMap: TemplateMap;
  views: string[];
}
export interface AppState {
  // meta-data
  currentlySelectedFile: string;
  columns: ColumnHeader[];
  metaColumns: ColumnHeader[];

  // spec configs
  spec: any;
  specCode: string;
  currentTheme: string;
  editorError: boolean;
  editMode: boolean;

  // GUI
  // currentTemplateInstance: SCATTERPLOT;
  // encodingMode: 'Scatterplot';
  currentTemplateInstance: Template | null;

  dataModalOpen: boolean;
  encodingMode: string;
  showProgrammaticMode: boolean;
  showSimpleDisplay: boolean;
  showGUIView: boolean;
  codeMode: string;
  editorFontSize: number;
  programModalOpen: boolean;

  // undo redo
  undoStack: UndoRedoStackItem[];
  redoStack: UndoRedoStackItem[];

  // view stuff
  views: string[];
  viewCatalog: {[x: string]: any};
  currentView: string;

  // template stuff

  templates: Template[];
  templateMap: TemplateMap;
  templateBuilderModalOpen: boolean;

  data: {[x: string]: any}[];
  originalData: {[x: string]: any}[];
  dataModification: string;
}

export interface ActionResponse {
  (state: AppState, payload: any): AppState;
}

export const blindSet = (key: string): ActionResponse => (state, payload): AppState =>
  produce(state, draftState => {
    /* eslint-disable @typescript-eslint/ban-ts-ignore*/
    // @ts-ignore
    draftState[key] = payload;
    /* eslint-enable @typescript-eslint/ban-ts-ignore*/
  });
export const toggle = (key: string): ActionResponse => (state): AppState =>
  produce(state, draftState => {
    /* eslint-disable @typescript-eslint/ban-ts-ignore*/
    // @ts-ignore
    draftState[key] = !state[key];
    /* eslint-enable @typescript-eslint/ban-ts-ignore*/
  });

// TODO undo this embarrasment (specifically the type messes)
const defaultEmpty = {
  $schema: 'https://vega.github.io/schema/vega-lite/v4.json',
  transform: [] as any[],
  mark: {type: 'point', tooltip: true},
  encoding: {},
};
export const EMPTY_SPEC = defaultEmpty;
export const DEFAULT_STATE: AppState = {
  // meta-data
  currentlySelectedFile: 'cars.json',
  columns: [],
  metaColumns: [],

  // spec configs
  spec: EMPTY_SPEC,
  specCode: stringify(EMPTY_SPEC),
  currentTheme: 'default',
  editorError: null,
  editMode: false,

  // GUI
  // currentTemplateInstance: SCATTERPLOT,
  // encodingMode: 'Scatterplot',
  currentTemplateInstance: null,

  dataModalOpen: false,
  encodingMode: 'grammer',
  showProgrammaticMode: true,
  showSimpleDisplay: false,
  showGUIView: true,
  codeMode: 'EXPORT TO JSON',
  editorFontSize: 10,
  programModalOpen: false,

  // undo redo
  undoStack: [],
  redoStack: [],

  // view stuff
  views: ['view1'],
  viewCatalog: {},
  currentView: 'view1',

  // template stuff

  templates: [],
  templateMap: {},
  templateBuilderModalOpen: false,

  data: [],
  originalData: [],
  dataModification: null,
};
