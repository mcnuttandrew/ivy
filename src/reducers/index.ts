import {createStore, combineReducers, applyMiddleware} from 'redux';
import thunk from 'redux-thunk';
// import {Spec} from 'vega-typings';

import {
  setEncodingParameter,
  setFacetingParameter,
  clearEncoding,
  changeMarkType,
  setNewSpec,
  setNewSpecCode,
  addToNextOpenSlot,
  coerceType,
  triggerRedo,
  triggerUndo,
  pushToUndoStack,
} from './modify-encodings';

import {createFilter, updateFilter, deleteFilter} from './filter-actions';
import {
  recieveData,
  recieveTypeInferences,
  changeSelectedFile,
} from './data-actions';
import {AppState, DEFAULT_STATE, ActionResponse} from './default-state';

// GUI ACTIONS
const changeGUIMode: ActionResponse = (state, payload) =>
  state.set('selectedGUIMode', payload);
const toggleDataModal: ActionResponse = state =>
  state.set('dataModalOpen', !state.get('dataModalOpen'));
const changeTheme: ActionResponse = (state, payload) =>
  state.set('currentTheme', payload);

const wrap = (func: ActionResponse, wrapper: any): ActionResponse => (
  state,
  payload,
) => wrapper(state, func(state, payload));

const actionFuncMap: {[val: string]: ActionResponse} = {
  // data modifications
  'change-selected-file': changeSelectedFile,
  'recieve-data': recieveData,
  'recieve-type-inferences': recieveTypeInferences,

  // encoding modifications
  'add-to-next-open-slot': wrap(addToNextOpenSlot, pushToUndoStack),
  'change-mark-type': wrap(changeMarkType, pushToUndoStack),
  'clear-encoding': wrap(clearEncoding, pushToUndoStack),
  'coerce-type': wrap(coerceType, pushToUndoStack),
  'set-encoding-param': wrap(setEncodingParameter, pushToUndoStack),
  'set-faceting-param': wrap(setFacetingParameter, pushToUndoStack),
  'set-new-encoding': wrap(setNewSpec, pushToUndoStack),
  'set-new-encoding-code': wrap(setNewSpecCode, pushToUndoStack),

  'trigger-redo': triggerRedo,
  'trigger-undo': triggerUndo,

  // filter modifications
  'create-filter': wrap(createFilter, pushToUndoStack),
  'delete-filter': wrap(deleteFilter, pushToUndoStack),
  'update-filter': wrap(updateFilter, pushToUndoStack),

  // gui modifications
  'change-gui-mode': changeGUIMode,
  'change-theme': changeTheme,
  'toggle-data-modal': toggleDataModal,
};
const NULL_ACTION: ActionResponse = state => state;

export default createStore(
  combineReducers({
    base: (
      state: AppState = DEFAULT_STATE,
      {type, payload}: {type: string, payload: any},
    ) => {
      return (actionFuncMap[type] || NULL_ACTION)(state, payload);
    },
  }),
  applyMiddleware(thunk),
);
