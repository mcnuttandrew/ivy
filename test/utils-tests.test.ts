import {
  setTemplateValues,
  fillTemplateMapWithDefaults,
  recieveTemplates,
} from '../src/reducers/template-actions';
import {DEFAULT_TEMPLATES} from '../src/constants/templates';
import SCATTERPLOT_TEMPLATE from '../src/constants/example-templates/scatterplot';
import PIECHART_TEMPLATE from '../src/constants/example-templates/pie-chart';

import {DEFAULT_STATE} from '../src/reducers/default-state';

test('#setTemplateValues', () => {
  const filledOutPieTemplate = setTemplateValues(PIECHART_TEMPLATE.code, {
    category: '"CATEGORY_TEST"',
    value: '"VALUE_TEST"',
    sortValues: 'true',
  });
  expect(filledOutPieTemplate).toMatchSnapshot();

  const filledOutScatterTemplate = setTemplateValues(
    SCATTERPLOT_TEMPLATE.code,
    {
      xDim: '"xDim_TEST"',
      yDim: '"yDim_TEST"',
    },
  );
  expect(filledOutScatterTemplate).toMatchSnapshot();
});

test('#fillTemplateMapWithDefaults', () => {
  const preparedState = recieveTemplates(DEFAULT_STATE, DEFAULT_TEMPLATES).set(
    'encodingMode',
    'scatterplot',
  );
  const newState = fillTemplateMapWithDefaults(preparedState);
  expect(newState.get('spec').toJS()).toMatchSnapshot();
  expect(newState.get('templateMap').toJS()).toMatchSnapshot();

  const nextState = fillTemplateMapWithDefaults(
    newState.set('encodingMode', 'pie chart'),
  );
  expect(nextState.get('spec').toJS()).toMatchSnapshot();
  expect(nextState.get('templateMap').toJS()).toMatchSnapshot();
});

// // UGH come back to this later
//
// // @ts-ignore
// import React from 'react';
// import {mount} from 'enzyme';
// import AppWrap from '../src/components/app-wrap';
//
// test('test mount', t => {
//   const $ = mount(<AppWrap />);
//   console.log($);
// });
