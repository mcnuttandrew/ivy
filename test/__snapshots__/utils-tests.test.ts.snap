// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fillTemplateMapWithDefaults 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
  "encoding": Object {
    "color": Object {
      "CONDITIONAL": Object {
        "false": null,
        "query": Object {
          "Color": "*",
        },
        "true": Object {
          "field": null,
          "type": "ordinal",
        },
      },
    },
    "x": Object {
      "field": null,
      "scale": Object {
        "zero": true,
      },
      "type": "quantitative",
    },
    "y": Object {
      "field": null,
      "scale": Object {
        "zero": true,
      },
      "type": "quantitative",
    },
  },
  "mark": Object {
    "color": Object {
      "CONDITIONAL": Object {
        "false": null,
        "query": Object {
          "Color": null,
        },
        "true": "steelblue",
      },
    },
    "size": 15,
    "tooltip": true,
    "type": "point",
  },
}
`;

exports[`#fillTemplateMapWithDefaults 2`] = `
Object {
  "Color": null,
  "OtherSettingsSection": null,
  "Radius": 15,
  "Single Color": "\\"steelblue\\"",
  "Zeroes": "true",
  "colorType": "\\"ordinal\\"",
  "xDim": null,
  "xType": "\\"quantitative\\"",
  "yDim": null,
  "yType": "\\"quantitative\\"",
}
`;

exports[`#fillTemplateMapWithDefaults 3`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega/v5.json",
  "autosize": "fit",
  "data": Array [
    Object {
      "name": "table",
      "transform": Array [
        Object {
          "as": Array [
            "pieAg",
          ],
          "fields": Array [
            null,
          ],
          "groupby": Array [
            null,
          ],
          "ops": Array [
            "mean",
          ],
          "type": "aggregate",
        },
        Object {
          "endAngle": 6.29,
          "field": "pieAg",
          "sort": true,
          "startAngle": 0,
          "type": "pie",
        },
        Object {
          "as": "tooltip",
          "expr": "datum.[category] + ': ' + datum.pieAg",
          "type": "formula",
        },
      ],
      "values": "myData",
    },
  ],
  "height": 200,
  "marks": Array [
    Object {
      "encode": Object {
        "enter": Object {
          "cornerRadius": Object {
            "value": 0,
          },
          "endAngle": Object {
            "field": "endAngle",
          },
          "fill": Object {
            "field": null,
            "scale": "color",
          },
          "innerRadius": Object {
            "value": 60,
          },
          "outerRadius": Object {
            "signal": "width / 2",
          },
          "startAngle": Object {
            "field": "startAngle",
          },
          "tooltip": Object {
            "field": "tooltip",
          },
          "x": Object {
            "signal": "width / 2",
          },
          "y": Object {
            "signal": "height / 2",
          },
        },
      },
      "from": Object {
        "data": "table",
      },
      "type": "arc",
    },
    Object {
      "encode": Object {
        "enter": Object {
          "align": Object {
            "value": "center",
          },
          "baseline": Object {
            "value": "middle",
          },
          "fill": Object {
            "value": "#000",
          },
          "radius": Object {
            "scale": "r",
            "value": 1.3,
          },
          "text": Object {
            "field": null,
          },
          "theta": Object {
            "signal": "(datum.startAngle + datum.endAngle)/2",
          },
          "x": Object {
            "field": Object {
              "group": "width",
            },
            "mult": 0.5,
          },
          "y": Object {
            "field": Object {
              "group": "height",
            },
            "mult": 0.5,
          },
        },
      },
      "from": Object {
        "data": "table",
      },
      "type": "text",
    },
  ],
  "padding": 50,
  "scales": Array [
    Object {
      "domain": Object {
        "data": "table",
        "field": null,
      },
      "name": "color",
      "range": Object {
        "scheme": "category20",
      },
      "type": "ordinal",
    },
    Object {
      "domain": Array [
        0,
        1,
      ],
      "name": "r",
      "range": Array [
        90,
        100,
      ],
      "type": "sqrt",
      "zero": false,
    },
  ],
  "width": 200,
}
`;

exports[`#fillTemplateMapWithDefaults 4`] = `
Object {
  "DonutChart": "60",
  "Sort": "true",
  "aggregate": "\\"mean\\"",
  "category": null,
  "colorScheme": "\\"category20\\"",
  "value": null,
}
`;

exports[`#fillTemplateMapWithDefaults 5`] = `
Object {
  "$schema": "hydra-data-table",
  "columns": Array [],
  "transforms": Array [],
}
`;

exports[`#fillTemplateMapWithDefaults 6`] = `
Object {
  "columns": Array [],
}
`;

exports[`#setTemplateValues 1`] = `
"{
  \\"$schema\\": \\"https://vega.github.io/schema/vega/v5.json\\",
  \\"width\\": 200,
  \\"height\\": 200,
  \\"padding\\": 50,
  \\"autosize\\": \\"fit\\",
  \\"data\\": [
    {
      \\"name\\": \\"table\\",
      \\"values\\": \\"myData\\",
      \\"transform\\": [
        {
          \\"type\\": \\"aggregate\\",
          \\"groupby\\": [\\"CATEGORY_TEST\\"],
          \\"fields\\": [\\"VALUE_TEST\\"],
          \\"ops\\": [\\"[aggregate]\\"],
          \\"as\\": [\\"pieAg\\"]
        },
        {
          \\"type\\": \\"pie\\",
          \\"field\\": \\"pieAg\\",
          \\"startAngle\\": 0,
          \\"endAngle\\": 6.29,
          \\"sort\\": \\"[Sort]\\"
        },
        {
          \\"type\\": \\"formula\\",
          \\"expr\\": \\"datum.CATEGORY_TEST + ': ' + datum.pieAg\\",
          \\"as\\": \\"tooltip\\"
        }
      ]
    }
  ],
  \\"scales\\": [
    {
      \\"name\\": \\"color\\",
      \\"type\\": \\"ordinal\\",
      \\"domain\\": {\\"data\\": \\"table\\", \\"field\\": \\"CATEGORY_TEST\\"},
      \\"range\\": {\\"scheme\\": \\"[colorScheme]\\"}
    },
    {
      \\"name\\": \\"r\\",
      \\"type\\": \\"sqrt\\",
      \\"domain\\": [0, 1],
      \\"zero\\": false,
      \\"range\\": [90, 100]
    }
  ],
  \\"marks\\": [
    {
      \\"type\\": \\"arc\\",
      \\"from\\": {\\"data\\": \\"table\\"},
      \\"encode\\": {
        \\"enter\\": {
          \\"fill\\": {\\"scale\\": \\"color\\", \\"field\\": \\"CATEGORY_TEST\\"},
          \\"x\\": {\\"signal\\": \\"width / 2\\"},
          \\"y\\": {\\"signal\\": \\"height / 2\\"},
          \\"startAngle\\": {\\"field\\": \\"startAngle\\"},
          \\"endAngle\\": {\\"field\\": \\"endAngle\\"},
          \\"innerRadius\\": {\\"value\\": \\"[DonutChart]\\"},
          \\"outerRadius\\": {\\"signal\\": \\"width / 2\\"},
          \\"cornerRadius\\": {\\"value\\": 0},
          \\"tooltip\\": {\\"field\\": \\"tooltip\\"}
        }
      }
    },
    {
      \\"type\\": \\"text\\",
      \\"from\\": {\\"data\\": \\"table\\"},
      \\"encode\\": {
        \\"enter\\": {
          \\"x\\": {\\"field\\": {\\"group\\": \\"width\\"}, \\"mult\\": 0.5},
          \\"y\\": {\\"field\\": {\\"group\\": \\"height\\"}, \\"mult\\": 0.5},
          \\"radius\\": {\\"scale\\": \\"r\\", \\"value\\": 1.3},
          \\"theta\\": {\\"signal\\": \\"(datum.startAngle + datum.endAngle)/2\\"},
          \\"fill\\": {\\"value\\": \\"#000\\"},
          \\"align\\": {\\"value\\": \\"center\\"},
          \\"baseline\\": {\\"value\\": \\"middle\\"},
          \\"text\\": {\\"field\\": \\"CATEGORY_TEST\\"}
        }
      }
    }
  ]
}"
`;

exports[`#setTemplateValues 2`] = `
"{
  \\"$schema\\": \\"https://vega.github.io/schema/vega-lite/v4.json\\",
  \\"mark\\": {
    \\"type\\": \\"point\\",
    \\"tooltip\\": true,
    \\"size\\": \\"[Radius]\\",
    \\"color\\": {
      \\"CONDITIONAL\\": {
        \\"true\\": \\"[Single Color]\\",
        \\"false\\": null,
        \\"query\\": {\\"Color\\": null}
      }
    }
  },
  \\"encoding\\": {
    \\"x\\": {\\"field\\": \\"xDim_TEST\\", \\"type\\": \\"[xType]\\", \\"scale\\": {\\"zero\\": \\"[Zeroes]\\"}},
    \\"y\\": {\\"field\\": \\"yDim_TEST\\", \\"type\\": \\"[yType]\\", \\"scale\\": {\\"zero\\": \\"[Zeroes]\\"}},
    \\"color\\": {
      \\"CONDITIONAL\\": {
        \\"query\\": {\\"Color\\": \\"*\\"},
        \\"true\\": {\\"field\\": \\"[Color]\\", \\"type\\": \\"[colorType]\\"},
        \\"false\\": null
      }
    }
  }
}"
`;

exports[`#setTemplateValues 3`] = `"{\\"$schema\\": \\"hydra-data-table\\", \\"transforms\\": [], \\"columns\\": [\\"\\\\\\"xDim_TEST\\\\\\"\\",\\"\\\\\\"yDim_TEST\\\\\\"\\"]}"`;
